docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,302 @@
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               'utc' => new UTCDateTime($value),
+               'tz' => $value->getTimezone()->getName(),
+           ]);
+       }
+   }
+
+This codec can now be leveraged by other codecs encode and decode dates. Let's return to the previous example and add a
+``createdAt`` field that contains the creation date. The modified encode and decode methods would look like this (other
+code omitted for brevity):
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       private DateTimeCodec $dateTimeCodec;
+
+       public function __construct()
+       {
+           $this->dateTimeCodec = new DateTimeCodec();
+       }
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+           $person->createdAt = $this->dateTimeCodec->decode($value->get('createdAt'));
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+               'createdAt' => $this->dateTimeCodec->encode($value->createdAt),
+           ]);
+       }
+   }
+
+Codec Libraries
+---------------
+
+If you have a number of codecs that you want to use in multiple places, you can create a codec library. A codec library
+contains a list of codecs and checks each codec if it supports a value. If it does, it will use that codec to encode or
+decode the given value. The following code snippet changes the PersonCodec to use a codec library instead of a
+hard-coded ``DateTimeCodec``:
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       private MongoDB\Codec\CodecLibrary $library;
+
+       public function __construct(private MongoDB\Codec\CodecLibrary $codecLibrary)
+       {
+       }
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+           $person->createdAt = $this->codecLibrary->decode($value->get('createdAt'));
```

Comment:
I think it's not possible, but the codecLibrary could be used as a service locator because you know which codec you need to use here. And from my previous comment, if you have a `City` property, you must use the `CityCodec`.

I'm thinking of the performance impact of having to loop through all the codecs for each field.